JDK-8289524: Add JFR JIT restart event

MBaesken opened the pull request:
The JIT compiler restarts (see restart_compiler in NMethodSweeper::sweep_code_cache) would be a helpful addition to the JFR events. Currently we log the JIT stop operations in JFR (EventCodeCacheFull) but no restart.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289524](https://bugs.openjdk.org/browse/JDK-8289524): Add JFR JIT restart event


### Reviewers
 * @parttimenerd (no known github.com user name / role) ⚠️ Review applies to [29665ecb](https://git.openjdk.org/jdk/pull/9334/files/29665ecb9f3a86077fcc564daacc8a5cd205568e)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [29665ecb](https://git.openjdk.org/jdk/pull/9334/files/29665ecb9f3a86077fcc564daacc8a5cd205568e)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9334/head:pull/9334` \
`$ git checkout pull/9334`

Update a local copy of the PR: \
`$ git checkout pull/9334` \
`$ git pull https://git.openjdk.org/jdk pull/9334/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9334`

View PR using the GUI difftool: \
`$ git pr show -t 9334`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9334.diff">https://git.openjdk.org/jdk/pull/9334.diff</a>

</details>

parttimenerd reviewed (approved):
Looks good and simple

MBaesken commented:
> Looks good and simple

Hi, thanks for the review ! May I have a second review ?

mgronlun commented:
Thanks for clarifying Erik, I forgot to mention the contentType.

MBaesken commented:
Hi Markus,  I think it would be possible and probably make sense to instead enhance the existing EventSweepCodeCache.
Would it be okay to move down the EventSweepCodeCache to the end of NMethodSweeper::sweep_code_cache()  ?
I think post_sweep_event  would get  2 more parameters, a boolean for JIT-restart and an integer freed_memory.
Are you fine with this ?
Are there some compatibility concerns when enhancing an existing event EventSweepCodeCache ?

mgronlun commented:
Thanks for considering. It is fine to move the EventSweepCodeCache event around, you can see that the constructor takes the UNTIMED value. This means that timestamping is handled explicitly, by setting the set_starttime() and set_endtime() fields For compatibility, we can think of it like we are creating a subclass with extended fields. Just append the additional fields to the event. As for the memory field declarations, you can take a peek at other events in metadata.xml, for example, this field declaration is from CodeCacheFull:

    <Field type="ulong" contentType="bytes" name="unallocatedCapacity" label="Unallocated" />

MBaesken commented:
Hi Markus, another question that came up while looking into this - the current SweepCodeCache event has a threshold of 100 ms set in both default.jfc and profile.jfc .  This was probably fine for existing usages.  But would we loose the JIT restart events sometimes in case we incorporate the Jit restart and frred memory into the current SweepCodeCache   event ?

mgronlun commented:
Can you explain a bit more about what "JIT restart" actually means?

MBaesken commented:
> Can you explain a bit more about what "JIT restart" actually means?

The comment at line 433 of sweeper.cpp and following is explaining it.
The JIT compilation had been stopped before and now, after sweeping freed potentially some memory,
JIT compilation  is re-enabled.
That's why also the  log.debug("restart compiler");  in the code.

mgronlun commented:
Ok, so there is a corresponding "compiler stopped", implicitly noted by firing EventCodeCacheFull?

mgronlun commented:
Hi again Mathias, based on your observation that this does not really align well with the threshold parameter for SweepCodeCache and is a complement event to EventCodeCacheFull, I believe your original suggestion is better (an instant event, only for "jit restart"). I was slightly tripped up with the concept of a  "JIT restart", because at this point the JIT is already stopped, as implied by EventCodeCacheFull. So "JIT restart" here contextually means, "JIT start", as in "start JITting code again". Do you know if we have an event that describes the CodeCache settings in terms of memory sizes so that the "freed memory" can be interpreted relative to it?

egahlin commented:
It would be good to add a sanity test.  See:
https://github.com/openjdk/jdk/blob/83a5d5996bca26b5f2e97b67f9bfd0a6ad110327/test/jdk/jdk/jfr/event/compiler/TestCodeSweeper.java#L181

The test in on the ProblemList.txt due to timeouts, but probably works most of the time. 

Also remove ""JitRestart" from TestLookForUntestedEvents.java

MBaesken commented:
> Yes, thank you, I took a look at those fields too. Unfortunately, there is no value that bears directly on "current in-use". The reserved and committed will not be updated. If there was a value that could reflect the current usage, even if it means adding it to both EventCodeCacheFull and the JIT restart, that would be perfect. EventCodeCacheFull because in-use is x, JIT restart because in-use is now y. There are some statistics-related properties in the CodeHeap, like for example heap->unallocated_capacity(); et al. Could some of those expose this running value perhaps?

Probably we would need  CodeCache::unallocated_capacity()  (contains a number of code heaps as far as I know)  to compare freed_memory to, because from what I see we iterate the whole CodeCache in NMethodSweeper::sweep_code_cache() .
But maybe someone with better knowledge about the CodeCache should comment here.

RealLucy commented:
To get an idea of how much memory was freed in relation to what's actually available, you could evaluate CodeCache:max_capacity(). This function returns the overall size of the code heap (across all segments). The performance impact is minimal and no locks are acquired.

MBaesken commented:
So is the common opinion to get back to a separate JIT start  event  (I think that naming is prefered over JIT restart, am I correct) ?
Additionally we add CodeCache:max_capacity() at both JIT start  and  EventCodeCacheFull ?

RealLucy commented:
> So is the common opinion to get back to a separate JIT start event (I think that naming is prefered over JIT restart, am I correct) ?
> Additionally we add CodeCache:max_capacity() at both JIT start and EventCodeCacheFull ?

I would be in favour of that.

mgronlun commented:
> So is the common opinion to get back to a separate JIT start event (I think that naming is prefered over JIT restart, am I correct) ? Additionally we add CodeCache:max_capacity() at both JIT start and EventCodeCacheFull ?

I think it is fine to use the term "JIT restart" because it is in use both in the code as well as in the output of log statements. Another reason would be that the first "JIT start" event would always be missing. It was only the meaning that got me a bit confused. Yes, a separate event "JIT is preferrable having no duration (startTime=false), no stack trace(stackTrace=false). The .jfc configs only need one element:

<setting name="enabled" control="compiler-enabled">true</setting>

RealLucy reviewed (approved):
Changes look good. 
Now you can relate the freed memory to what's available in total.

mgronlun commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -558,6 +558,11 @@
 
+  <Event name="JitRestart" category="Java Virtual Machine, Compiler" label="JIT restart" stackTrace="false" startTime="false" thread="true">
+    <Field type="int" name="freedMemory" label="Freed Memory" />
+    <Field type="ulong" name="codeCacheMaxCapacity" label="CodeCache maximum capacity" />
```
Is codeCacheMaxCapacity the current in-use size of the CodeCache, in bytes? It should have the contentType="bytes" in that case. Also "freedMemory" should have the same contentType.

RealLucy replied:
Yes, codeCacheMaxCapacity is given in bytes. It specifies the maximum size, not the current in-use size, of the CodeCache. It is the sum over all CodeHeap segments.

mgronlun commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -612,6 +617,7 @@
     <Field type="int" name="adaptorCount" label="Adaptors" />
     <Field type="ulong" contentType="bytes" name="unallocatedCapacity" label="Unallocated" />
     <Field type="int" name="fullCount" label="Full Count" />
+    <Field type="ulong" name="codeCacheMaxCapacity" label="CodeCache maximum capacity" />
```
contentType="bytes"

mgronlun commented on test/jdk/jdk/jfr/event/metadata/TestLookForUntestedEvents.java:
```diff
@@ -54,7 +54,7 @@ public class TestLookForUntestedEvents {
     private static final Set<String> hardToTestEvents = new HashSet<>(
         Arrays.asList(
-            "DataLoss", "IntFlag", "ReservedStackActivation",
+            "DataLoss", "IntFlag", "ReservedStackActivation", "JitRestart",
```
Can we create a test? There is an existing test for CodeCacheFull, maybe derive from it to create a test also for the JitRestart event?

egahlin commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -558,6 +558,11 @@
     <Field type="ulong" contentType="bytes" name="used" label="Used" />
   </Event>
 
+  <Event name="JitRestart" category="Java Virtual Machine, Compiler" label="JIT restart" stackTrace="false" startTime="false" thread="true">
```
"JITRestart" . The convention for other events have been to use capital laters for acronyms. 

"JIT restart" -> "JIT Restart"

egahlin commented on src/hotspot/share/code/codeCache.cpp:
```diff
@@ -1366,6 +1366,7 @@ void CodeCache::report_codemem_full(int code_blob_type, bool print) {
     event.set_adaptorCount(heap->adapter_count());
     event.set_unallocatedCapacity(heap->unallocated_capacity());
     event.set_fullCount(heap->full_count());
+    event.set_codeCacheMaxCapacity(CodeCache::max_capacity());
```
Add test of field in TestCodeCacheFull

egahlin commented on src/hotspot/share/runtime/sweeper.cpp:
```diff
@@ -442,6 +437,15 @@ void NMethodSweeper::sweep_code_cache() {
     CompileBroker::set_should_compile_new_jobs(CompileBroker::run_compilation);
     log.debug("restart compiler");
     log_sweep("restart_compiler");
+    EventJitRestart event;
```
It would be good to have a sanity test of the event. 

If the event can't be provoked reliably, retry or accept as OK.

MBaesken commented:
Hi , I adjusted the contentType as suggested, adjusted the label  and TestCodeCacheFull.
Regarding triggering the JitRestart event , the  suggested TestCodeCacheFull looks like a good start . So I  added to allocateCodeBlob (this triggers the CodeCacheFull event)  a  corresponding freeCodeBlob call .
Unfortunately this did not work in practise and doesn't show the JitRestart event.  There could be some delay to the restarting unfortunately.

MBaesken commented:
I added a test after some playing around with the WhiteBox functionality.

coleenp commented on test/jdk/jdk/jfr/event/compiler/TestJitRestart.java:
```diff
@@ -0,0 +1,104 @@
+import jdk.test.lib.Asserts;
+import jdk.test.lib.jfr.EventNames;
+import jdk.test.lib.jfr.Events;
+import sun.hotspot.WhiteBox;
```
This package has been removed. Please use jdk.test.whitebox.WhiteBox.

MBaesken replied:
Hi Coleen thanks for the advice.  After switching to the new package  jdk.test.whitebox.WhiteBox  I get 
java.lang.UnsatisfiedLinkError: 'void sun.hotspot.WhiteBox.registerNatives()'
Do I need to do more than just renaming the package ?

MBaesken replied:
Had to change BlobType import  too, this one is also in a new package.

RealLucy reviewed (approved):
With all the fine-tuning, changes look even better now. And there is a test now so we get alerted if anything breaks. Thanks!

MBaesken commented:
/integrate
